src/Generator/Parsing.fs

```diff
@@ -0,0 +1,199 @@
+            let res =
+                { groupName = group
+                  cases = cases }
+            return! Some res
+        }
+    )
+
+let paramsTypeRegex = new Regex("<param(.*?(group=\"(?<group>.*?)\").*?|.*?)>(?<f>.*?)(<ptype>(?<t>.*?)<\/ptype>(?<b>.*?))?(<name>.+?<\/name>)<\/param>")
+let protoTypeRegex = new Regex("<proto(.*?(group=\"(?<group>.*?)\").*?|.*?)>(?<f>.*?)(<ptype>(?<t>.*?)<\/ptype>(?<b>.*?))?(<name>.+?<\/name>)<\/proto>")
+let getGroupValue (group: string) (matching: Match) =
+    matching.Groups.[group].Value
+
+let getRawElementDataWithoutLineBreaks (v: Xml.Linq.XElement) =
+    v.ToString().Replace("\n", "").Replace("\t", "").Replace("\r", "")
+
+let extractTypeFromPtype (param: OpenGL_Specification.Param) =
+    let str = getRawElementDataWithoutLineBreaks param.XElement
+    let res = paramsTypeRegex.Match str
+    let groupName =
+        let str = getGroupValue "group" res
+        if str |> String.IsNullOrWhiteSpace then None
+        else Some str
+    let ty =
+        getGroupValue "f" res +
+        getGroupValue "t" res +
+        getGroupValue "b" res
+    groupName, ty
+
+let extractTypeFromProto (proto: OpenGL_Specification.Proto) =
+    let str = getRawElementDataWithoutLineBreaks proto.XElement
+    let res = protoTypeRegex.Match str
+    let groupName =
+        let str = getGroupValue "group" res
+        if str |> String.IsNullOrWhiteSpace then None
+        else Some str
+    let ty =
+        getGroupValue "f" res +
+        getGroupValue "t" res +
+        getGroupValue "b" res
+    groupName, ty
+
+let getFunctions (spec: OpenGL_Specification.Registry) =
+    spec.Commands.Commands
+    |> Array.Parallel.map(fun cmd ->
+        //let isInvalid =
+        //    cmd.Params |> Array.exists(fun p -> p.Ptype |> Option.isNone)
+        //    || Option.isNone cmd.Proto.Ptype 
+        //if isInvalid then
+        //    printfn "Function specification %A is invalid" cmd.Proto.Name
+        //    cmd.Params |> Seq.iter (fun p -> printfn "%A" p.XElement.Value)
+        //    None
+        //else
+            let funcName = cmd.Proto.Name
+            let parameters =
+                cmd.Params
+                |> Array.Parallel.map(fun p ->
+                    let group, ty = extractTypeFromPtype p
+                    if String.IsNullOrWhiteSpace ty then
+                        let str = p.XElement.ToString()
+                        printfn "failed parsing %A, value: %s" (funcName) str
+                    { paramName = p.Name
+                      paramType = looseType ty group }
+                )
+            let group, ty = extractTypeFromProto cmd.Proto
+            let ret =
+                { funcName = funcName 
+                  parameters = parameters
+                  retType = looseType ty group }
+            ret
+    )
+
+let getExtensions (spec: OpenGL_Specification.Registry) =
+    spec.Extensions
+    |> Array.Parallel.map(fun extension ->
+        let functions, enums =
```

Comment:
why are functions & enums defined twice?

Reply:
Because i dont want to pipe it into a lambda